File: 202012/20201211/q783.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    def minDiffInBST(self, root: TreeNode) -> int:
        stack, prev, minV = [], None, None
        while stack or root:
            while root:
                stack.append(root)
                root = root.left

            node = stack.pop(-1)
            if prev:
                if minV:
                    minV = min(minV, node.val - prev)
                else:
                    minV = node.val - prev

            prev = node.val

            if node.right:
                root = node.right

        return minV

File: 202104/20210413/q783.java
import java.util.LinkedList;

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}
class Solution {
    public int minDiffInBST(TreeNode root) {
        LinkedList<TreeNode> list = new LinkedList<>();
        TreeNode pre = null;
        int result = (int) Math.pow(10, 6);
        while (null != root || !list.isEmpty()) {
            while (null != root) {
                list.add(0, root);
                root = root.left;
            }
            TreeNode pop = list.pop();
            if (null == pre) {
                pre = pop;
            } else {
                result = Math.min(result, pop.val - pre.val);
                pre = pop;
            }
            if (null != pop.right) {
                root = pop.right;
            }
        }
        return result;
    }
}